File: packages/php/src/Seguranca/Ataques/AtaqueBateriaEstatistica.php
<?php

namespace Application\Seguranca\Ataques;

use Application\Seguranca\AlvoCriptografico;
use Application\Seguranca\AtaqueInterface;
use Application\Seguranca\ResultadoAtaque;
use Application\Seguranca\SkipAtaqueException;

/**
 * Bateria estatística no keystream (inspirada em NIST SP800-22): frequência
 * global (monobit), frequência por posição de bit, teste de runs e frequência
 * por bloco. Diferente do qui-quadrado de bytes, aqui o foco é o nível de BIT
 * e a estrutura de sequência - vieses que a contagem de bytes pode mascarar.
 *
 * Usa z-scores com limite conservador (4 desvios) em vez de p-valores exatos,
 * pra não depender de funções numéricas especiais.
 */
class AtaqueBateriaEstatistica implements AtaqueInterface
{
    public function __construct(private int $tamanho = 16384, private float $zLimite = 4.0)
    {
    }

    public function nome(): string
    {
        return 'Bateria estatística de bits (monobit/runs/blocos)';
    }

    public function executar(AlvoCriptografico $alvo): ResultadoAtaque
    {
        $ks = $alvo->gerarKeystreamBruto($alvo->chaveDeTeste(), random_bytes($alvo->tamanhoIv()), 'enc', $this->tamanho);
        if ($ks === null) {
            throw new SkipAtaqueException('Alvo não expõe gerarKeystreamBruto.');
        }

        $bits = $this->paraBits($ks);
        $n = count($bits);
        $problemas = [];
        $dados = [];

        // 1) Monobit: soma +1/-1
        $soma = 0;
        foreach ($bits as $b) {
            $soma += $b === 1 ? 1 : -1;
        }
        $zMonobit = abs($soma) / sqrt($n);
        $dados['monobit_z'] = $zMonobit;
        if ($zMonobit > $this->zLimite) {
            $problemas[] = sprintf('monobit z=%.2f', $zMonobit);
        }

        // 2) Frequência por posição de bit
        $unsPorPos = array_fill(0, 8, 0);
        $contaPorPos = array_fill(0, 8, 0);
        foreach (str_split($ks) as $c) {
            $byte = ord($c);
            for ($b = 0; $b < 8; $b++) {
                if ((($byte >> $b) & 1) === 1) {
                    $unsPorPos[$b]++;
                }
                $contaPorPos[$b]++;
            }
        }
        $posViesadas = [];
        foreach ($unsPorPos as $b => $uns) {
            $frac = $uns / $contaPorPos[$b];
            $z = abs($frac - 0.5) / (0.5 / sqrt($contaPorPos[$b]));
            if ($z > $this->zLimite) {
                $posViesadas[$b] = $frac;
            }
        }
        $dados['bit_posicao_viesadas'] = $posViesadas;
        if (!empty($posViesadas)) {
            $problemas[] = 'viés por posição de bit: ' . implode(', ', array_keys($posViesadas));
        }

        // 3) Runs test
        $pi = array_sum($bits) / $n;
        if (abs($pi - 0.5) < 2 / sqrt($n)) {
            $runs = 1;
            for ($i = 1; $i < $n; $i++) {
                if ($bits[$i] !== $bits[$i - 1]) {
                    $runs++;
                }
            }
            $esperado = 2 * $n * $pi * (1 - $pi);
            $desvio = 2 * sqrt(2 * $n) * $pi * (1 - $pi);
            $zRuns = abs($runs - $esperado) / $desvio;
            $dados['runs_z'] = $zRuns;
            if ($zRuns > $this->zLimite) {
                $problemas[] = sprintf('runs z=%.2f', $zRuns);
            }
        }

        // 4) Frequência por bloco (M = 128 bits)
        $m = 128;
        $numBlocos = intdiv($n, $m);
        if ($numBlocos > 0) {
            $chi = 0.0;
            for ($i = 0; $i < $numBlocos; $i++) {
                $uns = 0;
                for ($j = 0; $j < $m; $j++) {
                    $uns += $bits[$i * $m + $j];
                }
                $chi += (($uns / $m) - 0.5) ** 2;
            }
            $chi *= 4 * $m;
            $zBlocos = ($chi - $numBlocos) / sqrt(2 * $numBlocos);
            $dados['blocos_chi2'] = $chi;
            $dados['blocos_z'] = $zBlocos;
            if ($zBlocos > $this->zLimite) {
                $problemas[] = sprintf('frequência por bloco chi2=%.1f', $chi);
            }
        }

        if (!empty($problemas)) {
            return new ResultadoAtaque($this->nome(), true, 'media', implode('; ', $problemas), $dados);
        }

        return new ResultadoAtaque(
            $this->nome(),
            false,
            'info',
            sprintf(
                'monobit z=%.2f, runs z=%.2f, blocos chi2=%.1f - todos abaixo do limite z=%.0f',
                $zMonobit,
                $dados['runs_z'] ?? 0.0,
                $dados['blocos_chi2'] ?? 0.0,
                $this->zLimite,
            ),
            $dados,
        );
    }

    /** @return list<int> */
    private function paraBits(string $bytes): array
    {
        $bits = [];
        for ($i = 0; $i < strlen($bytes); $i++) {
            $byte = ord($bytes[$i]);
            for ($b = 7; $b >= 0; $b--) {
                $bits[] = ($byte >> $b) & 1;
            }
        }
        return $bits;
    }
}

File: packages/rust/src/seguranca/ataques/ataque_bateria_estatistica.rs
//! Bateria estatística no keystream (inspirada em NIST SP800-22): frequência
//! global (monobit), frequência por posição de bit, teste de runs e frequência
//! por bloco. Diferente do qui-quadrado de bytes, aqui o foco é o nível de BIT
//! e a estrutura de sequência - vieses que a contagem de bytes pode mascarar.
//!
//! Usa z-scores com limite conservador (4 desvios) em vez de p-valores exatos,
//! pra não depender de funções numéricas especiais.

use crate::seguranca::alvo_criptografico::AlvoCriptografico;
use crate::seguranca::ataque::{Ataque, ErroAtaque};
use crate::seguranca::resultado_ataque::{ResultadoAtaque, Severidade};
use crate::seguranca::util::random_bytes;

pub struct AtaqueBateriaEstatistica {
    pub tamanho: usize,
    pub z_limite: f64,
}

impl AtaqueBateriaEstatistica {
    pub fn new() -> Self {
        AtaqueBateriaEstatistica {
            tamanho: 16384,
            z_limite: 4.0,
        }
    }
}

impl Default for AtaqueBateriaEstatistica {
    fn default() -> Self {
        Self::new()
    }
}

impl AtaqueBateriaEstatistica {
    fn para_bits(bytes: &[u8]) -> Vec<u8> {
        let mut bits: Vec<u8> = Vec::with_capacity(bytes.len() * 8);
        for &byte in bytes {
            for b in (0..8).rev() {
                bits.push((byte >> b) & 1);
            }
        }
        bits
    }
}

impl Ataque for AtaqueBateriaEstatistica {
    fn nome(&self) -> String {
        "Bateria estatística de bits (monobit/runs/blocos)".to_string()
    }

    fn executar(&self, alvo: &dyn AlvoCriptografico) -> Result<ResultadoAtaque, ErroAtaque> {
        let ks = alvo.gerar_keystream_bruto(
            alvo.chave_de_teste().as_bytes(),
            &random_bytes(alvo.tamanho_iv()),
            "enc",
            self.tamanho,
        );

        let bits = Self::para_bits(&ks);
        let n = bits.len();
        let mut problemas: Vec<String> = Vec::new();

        // 1) Monobit: soma +1/-1
        let mut soma: i64 = 0;
        for &b in &bits {
            soma += if b == 1 { 1 } else { -1 };
        }
        let z_monobit = soma.unsigned_abs() as f64 / (n as f64).sqrt();
        if z_monobit > self.z_limite {
            problemas.push(format!("monobit z={:.2}", z_monobit));
        }

        // 2) Frequência por posição de bit
        let mut uns_por_pos = [0usize; 8];
        let mut conta_por_pos = [0usize; 8];
        for &byte in &ks {
            for b in 0..8 {
                if ((byte >> b) & 1) == 1 {
                    uns_por_pos[b] += 1;
                }
                conta_por_pos[b] += 1;
            }
        }
        let mut pos_viesadas: Vec<usize> = Vec::new();
        for b in 0..8 {
            let frac = uns_por_pos[b] as f64 / conta_por_pos[b] as f64;
            let z = (frac - 0.5).abs() / (0.5 / (conta_por_pos[b] as f64).sqrt());
            if z > self.z_limite {
                pos_viesadas.push(b);
            }
        }
        if !pos_viesadas.is_empty() {
            let lista: Vec<String> = pos_viesadas.iter().map(|b| b.to_string()).collect();
            problemas.push("viés por posição de bit: ".to_string() + &lista.join(", "));
        }

        // 3) Runs test
        let mut runs_z: Option<f64> = None;
        let pi = bits.iter().map(|&b| b as usize).sum::<usize>() as f64 / n as f64;
        if (pi - 0.5).abs() < 2.0 / (n as f64).sqrt() {
            let mut runs = 1usize;
            for i in 1..n {
                if bits[i] != bits[i - 1] {
                    runs += 1;
                }
            }
            let esperado = 2.0 * n as f64 * pi * (1.0 - pi);
            let desvio = 2.0 * (2.0 * n as f64).sqrt() * pi * (1.0 - pi);
            let z_runs = (runs as f64 - esperado).abs() / desvio;
            runs_z = Some(z_runs);
            if z_runs > self.z_limite {
                problemas.push(format!("runs z={:.2}", z_runs));
            }
        }

        // 4) Frequência por bloco (M = 128 bits)
        let mut blocos_chi2: Option<f64> = None;
        let m = 128usize;
        let num_blocos = n / m;
        if num_blocos > 0 {
            let mut chi = 0.0f64;
            for i in 0..num_blocos {
                let mut uns = 0usize;
                for j in 0..m {
                    uns += bits[i * m + j] as usize;
                }
                chi += (uns as f64 / m as f64 - 0.5).powi(2);
            }
            chi *= 4.0 * m as f64;
            let z_blocos = (chi - num_blocos as f64) / (2.0 * num_blocos as f64).sqrt();
            blocos_chi2 = Some(chi);
            if z_blocos > self.z_limite {
                problemas.push(format!("frequência por bloco chi2={:.1}", chi));
            }
        }

        if !problemas.is_empty() {
            return Ok(ResultadoAtaque::new(
                &self.nome(),
                true,
                Severidade::Media,
                &problemas.join("; "),
            ));
        }

        Ok(ResultadoAtaque::new(
            &self.nome(),
            false,
            Severidade::Info,
            &format!(
                "monobit z={:.2}, runs z={:.2}, blocos chi2={:.1} - todos abaixo do limite z={:.0}",
                z_monobit,
                runs_z.unwrap_or(0.0),
                blocos_chi2.unwrap_or(0.0),
                self.z_limite
            ),
        ))
    }
}
